File: python/openocr/webhooks.py
"""OpenOCR webhook utilities — registration + signature verification."""

from __future__ import annotations

import hashlib
import hmac
import json
import time
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class WebhookEvent:
    """A parsed webhook event payload.

    Attributes:
        event_type: Event type (e.g. "job.complete", "job.failed").
        job_id: The OCR job ID.
        request_id: The OCR request ID.
        status: Job status ("succeeded" or "failed").
        engine: Engine ID used for processing.
        total_pages: Total pages in the document.
        extracted_text: Full extracted text (on success).
        error: Error message (on failure).
        timestamp: Unix timestamp of the event.
        usage: Usage/cost details.
    """
    event_type: str
    job_id: str
    request_id: str
    status: str
    engine: str
    total_pages: int = 0
    extracted_text: Optional[str] = None
    error: Optional[str] = None
    timestamp: float = 0.0
    usage: Optional[dict] = None

    @classmethod
    def from_dict(cls, data: dict) -> "WebhookEvent":
        """Parse a webhook event from a dictionary (JSON body)."""
        return cls(
            event_type=data.get("event_type", data.get("event", "")),
            job_id=data.get("job_id", ""),
            request_id=data.get("request_id", ""),
            status=data.get("status", ""),
            engine=data.get("engine", ""),
            total_pages=data.get("total_pages", 0),
            extracted_text=data.get("extracted_text"),
            error=data.get("error"),
            timestamp=data.get("timestamp", 0.0),
            usage=data.get("usage"),
        )

    @classmethod
    def from_json(cls, body: str | bytes) -> "WebhookEvent":
        """Parse a webhook event from a JSON string or bytes."""
        if isinstance(body, bytes):
            body = body.decode("utf-8")
        return cls.from_dict(json.loads(body))


def verify_signature(
    payload: str | bytes,
    signature: str,
    secret: str,
    *,
    tolerance_seconds: int = 300,
) -> bool:
    """Verify an OpenOCR webhook signature.

    The signature format is: ``t=<timestamp>,v1=<hex-hmac-sha256>``

    Verification:
    1. Parse the timestamp and HMAC from the signature header.
    2. Check the timestamp is within tolerance (default: 5 minutes).
    3. Compute ``HMAC-SHA256(secret, "<timestamp>.<payload>")``
    4. Compare against the provided HMAC using constant-time comparison.

    Args:
        payload: The raw request body (string or bytes).
        signature: The ``X-OpenOCR-Signature`` header value.
        secret: Your webhook signing secret.
        tolerance_seconds: Max age of the signature in seconds (default: 300).

    Returns:
        ``True`` if the signature is valid, ``False`` otherwise.

    Raises:
        ValueError: If the signature format is invalid.
    """
    if isinstance(payload, bytes):
        payload = payload.decode("utf-8")

    # Parse signature: t=<timestamp>,v1=<hmac>
    parts = {}
    for part in signature.split(","):
        if "=" in part:
            key, _, value = part.partition("=")
            parts[key.strip()] = value.strip()

    timestamp_str = parts.get("t")
    provided_hmac = parts.get("v1")

    if not timestamp_str or not provided_hmac:
        raise ValueError(
            f"Invalid signature format. Expected 't=<timestamp>,v1=<hmac>', got: {signature!r}"
        )

    # Check timestamp tolerance
    try:
        timestamp = int(timestamp_str)
    except ValueError:
        raise ValueError(f"Invalid timestamp in signature: {timestamp_str!r}")

    now = int(time.time())
    if abs(now - timestamp) > tolerance_seconds:
        return False

    # Compute expected HMAC
    signed_payload = f"{timestamp}.{payload}"
    expected_hmac = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    # Constant-time comparison
    return hmac.compare_digest(expected_hmac, provided_hmac)


def construct_signature(payload: str | bytes, secret: str, timestamp: Optional[int] = None) -> str:
    """Construct an OpenOCR webhook signature (for testing).

    Args:
        payload: The request body.
        secret: The webhook signing secret.
        timestamp: Optional unix timestamp (defaults to now).

    Returns:
        Signature string in format ``t=<timestamp>,v1=<hmac>``.
    """
    if isinstance(payload, bytes):
        payload = payload.decode("utf-8")

    if timestamp is None:
        timestamp = int(time.time())

    signed_payload = f"{timestamp}.{payload}"
    sig = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return f"t={timestamp},v1={sig}"

File: typescript/src/webhooks.ts
import { createHmac, timingSafeEqual } from "crypto";

/** A parsed webhook event payload. */
export interface WebhookEvent {
  eventType: string;
  jobId: string;
  requestId: string;
  status: string;
  engine: string;
  totalPages: number;
  extractedText?: string;
  error?: string;
  timestamp: number;
  usage?: Record<string, unknown>;
}

/**
 * Parse a webhook event from a JSON body.
 */
export function parseWebhookEvent(body: string | Record<string, unknown>): WebhookEvent {
  const data = typeof body === "string" ? JSON.parse(body) : body;
  return {
    eventType: data.event_type ?? data.event ?? "",
    jobId: data.job_id ?? "",
    requestId: data.request_id ?? "",
    status: data.status ?? "",
    engine: data.engine ?? "",
    totalPages: data.total_pages ?? 0,
    extractedText: data.extracted_text,
    error: data.error,
    timestamp: data.timestamp ?? 0,
    usage: data.usage,
  };
}

/**
 * Verify an OpenOCR webhook signature.
 *
 * The signature format is: `t=<timestamp>,v1=<hex-hmac-sha256>`
 *
 * Verification:
 * 1. Parse the timestamp and HMAC from the signature header.
 * 2. Check the timestamp is within tolerance (default: 5 minutes).
 * 3. Compute `HMAC-SHA256(secret, "<timestamp>.<payload>")`
 * 4. Compare against the provided HMAC using constant-time comparison.
 *
 * @param payload - The raw request body as a string.
 * @param signature - The `X-OpenOCR-Signature` header value.
 * @param secret - Your webhook signing secret.
 * @param toleranceSeconds - Max age of the signature in seconds (default: 300).
 * @returns `true` if the signature is valid.
 * @throws Error if the signature format is invalid.
 */
export function verifySignature(
  payload: string,
  signature: string,
  secret: string,
  toleranceSeconds: number = 300
): boolean {
  // Parse signature: t=<timestamp>,v1=<hmac>
  const parts: Record<string, string> = {};
  for (const part of signature.split(",")) {
    const eqIdx = part.indexOf("=");
    if (eqIdx !== -1) {
      parts[part.slice(0, eqIdx).trim()] = part.slice(eqIdx + 1).trim();
    }
  }

  const timestampStr = parts.t;
  const providedHmac = parts.v1;

  if (!timestampStr || !providedHmac) {
    throw new Error(
      `Invalid signature format. Expected 't=<timestamp>,v1=<hmac>', got: '${signature}'`
    );
  }

  const timestamp = parseInt(timestampStr, 10);
  if (isNaN(timestamp)) {
    throw new Error(`Invalid timestamp in signature: '${timestampStr}'`);
  }

  // Check timestamp tolerance
  const now = Math.floor(Date.now() / 1000);
  if (Math.abs(now - timestamp) > toleranceSeconds) {
    return false;
  }

  // Compute expected HMAC
  const signedPayload = `${timestamp}.${payload}`;
  const expectedHmac = createHmac("sha256", secret).update(signedPayload).digest("hex");

  // Constant-time comparison
  const a = Buffer.from(expectedHmac, "utf-8");
  const b = Buffer.from(providedHmac, "utf-8");
  if (a.length !== b.length) return false;
  return timingSafeEqual(a, b);
}

/**
 * Construct an OpenOCR webhook signature (for testing).
 *
 * @param payload - The request body.
 * @param secret - The webhook signing secret.
 * @param timestamp - Optional unix timestamp (defaults to now).
 * @returns Signature string in format `t=<timestamp>,v1=<hmac>`.
 */
export function constructSignature(
  payload: string,
  secret: string,
  timestamp?: number
): string {
  const ts = timestamp ?? Math.floor(Date.now() / 1000);
  const signedPayload = `${ts}.${payload}`;
  const sig = createHmac("sha256", secret).update(signedPayload).digest("hex");
  return `t=${ts},v1=${sig}`;
}
